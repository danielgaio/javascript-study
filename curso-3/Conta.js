export class Conta{
  constructor(saldoInicial, cliente, agencia) {
    this._saldo = saldoInicial;
    this._cliente = cliente;
    this._agencia = agencia;
  }

  set cliente(novoValor) {
    if (novoValor instanceof Cliente) {
      this._cliente = novoValor;
    }
  }

  get cliente() {
    return this._cliente;
  }

  get saldo() {
    return this._saldo;
  }
  
  sacar(valor) {
    let taxa = 1;
    const valorSacado = taxa * valor;
    if (this._saldo >= valorSacado) {
      this._saldo -= valorSacado;
      return valorSacado;
    } else {
      console.log("Não há saldo para sacar")
    }
  }

  depositar(valor) {
    // Técnica early returning: analisar casos de erro no inicio
    if (valor <= 0) return "Erro: Deposito de valor zero ou negativo";

    this._saldo += valor;
    console.log("Depósito de %s para %s", valor, this._cliente.nome);
  }

  exibirSaldo() {
    console.log("Saldo atual de %s: %s",this._cliente.nome, this._saldo);
  }

  // variáveis primitivas (valor) chegam na função como cópias, enquanto que
  // em objetos (conta) chega o original. Então alterar o objeto neste escopo
  // vai altera-lo no escopo externo, já com a variável primitiva isso não
  // acontece. Objetos são passados por referência.
  transferir(valor, conta) {
    const valorSacado = this.sacar(valor);
    conta.depositar(valorSacado);
  }
}
