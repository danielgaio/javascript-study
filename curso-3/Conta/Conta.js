// classe abstrata: uma classe que não pode ser instanciada diretamente
export class Conta{
  constructor(saldoInicial, cliente, agencia) {
    if (this.constructor == Conta) {
      throw new Error("Objeto do tipo conta não deve ser instanciado diretamente, pois essa é uma classe abstrata")
    } 
    this._saldo = saldoInicial;
    this._cliente = cliente;
    this._agencia = agencia;
  }

  set cliente(novoValor) {
    if (novoValor instanceof Cliente) {
      this._cliente = novoValor;
    }
  }

  get cliente() {
    return this._cliente;
  }

  get saldo() {
    return this._saldo;
  }
  
  // metodo abstrato
  sacar(valor) {
    throw new Error("O método sacar da conta é abstrato e deve ser sobreescrito");
  }

  _sacar(valor, taxa) {
    const valorSacado = taxa * valor;
    if (this._saldo >= valorSacado) {
      this._saldo -= valorSacado;
      return valorSacado;
    } else {
      return 0;
    }
  }

  depositar(valor) {
    // Técnica early returning: analisar casos de erro no inicio
    if (valor <= 0) return "Erro: Deposito de valor zero ou negativo";

    this._saldo += valor;
    console.log("Depósito de %s para %s", valor, this._cliente.nome);
  }

  exibirSaldo() {
    console.log("Saldo atual de %s: %s",this._cliente.nome, this._saldo);
  }

  // variáveis primitivas (valor) chegam na função como cópias, enquanto que
  // em objetos (conta) chega o original. Então alterar o objeto neste escopo
  // vai altera-lo no escopo externo, já com a variável primitiva isso não
  // acontece. Objetos são passados por referência.
  transferir(valor, conta) {
    const valorSacado = this.sacar(valor);
    conta.depositar(valorSacado);
  }
}
