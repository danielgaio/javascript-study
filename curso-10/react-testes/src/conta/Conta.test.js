import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import Conta from "./Conta";

describe("Componente de conta", () => {
  it("Snapshot do componente de conta", () => {
    const { container } = render(<Conta />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("Exibir o saldo da conta como valor monetário", () => {
    // primeiro renderizar o componente que quero testar
    render(<Conta saldo={1000} />);
    const saldo = screen.getByTestId("saldo-conta");
    expect(saldo.textContent).toBe("R$ 1000");
  });

  it("Chama a função de realizar transação, quando o botão é clicado", () => {
    // primeiro renderizar o componente a ser testado
    const funcaoRealizarTransacao = jest.fn(); // função mock, de simulação
    render(<Conta saldo={1000} realizarTransacao={funcaoRealizarTransacao} />);
    fireEvent.click(screen.getByText("Realizar operação"));
    expect(funcaoRealizarTransacao).toHaveBeenCalled();
  });
});
