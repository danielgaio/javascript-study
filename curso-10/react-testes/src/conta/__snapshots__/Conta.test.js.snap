// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente de conta Snapshot do componente de conta 1`] = `
<div
  class="Conta-header"
>
  <h2>
    Conta
  </h2>
  <p>
    Saldo:
     
    <span
      class="Saldo-valor"
      data-testid="saldo-conta"
    >
      R$ 0
    </span>
  </p>
  <form>
    <div>
      <label>
        Depósito
        <input
          data-testid="transacao"
          name="transacao"
          type="radio"
          value="deposito"
        />
      </label>
    </div>
    <div>
      <label>
        Saque
        <input
          data-testid="transacao"
          name="transacao"
          type="radio"
          value="saque"
        />
      </label>
    </div>
    <label>
      Valor:
    </label>
    <input
      data-testid="valor"
      name="valor"
      type="text"
      value="0"
    />
    <div>
      <button
        type="submit"
      >
        Realizar operação
      </button>
    </div>
  </form>
</div>
`;
